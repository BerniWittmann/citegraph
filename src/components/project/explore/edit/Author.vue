<template>
  <div class="project-explore-single-author-edit">
    <v-card class="pa-4">
      <v-card-title class="headline">
        <h2 class="display-5" v-once>{{ author.firstName }} {{ author.lastName }}</h2>
      </v-card-title>
      <v-container>
        <v-form ref="form" @submit.prevent="submit" class="mt-12" v-model="valid">
          <v-row justify="center">
            <v-col lg="6">
              <v-text-field
                :label="$t('project.explore.edit.author.first_name')"
                v-model="firstName"
                :rules="nameRules"
                required
                autofocus/>
            </v-col>
          </v-row>
          <v-row justify="center">
            <v-col lg="6">
              <v-text-field
                :label="$t('project.explore.edit.author.last_name')"
                v-model="lastName"
                :rules="nameRules"
                required
                autofocus/>
            </v-col>
          </v-row>
          <v-row class="my-10" justify="space-between">
            <v-col cols="auto">
              <v-btn rounded large color="warning" @click="back">
                <v-icon>mdi-menu-left</v-icon>
                {{ $t('project.explore.edit.back') }}
              </v-btn>
            </v-col>

            <v-col cols="auto">
              <v-btn
                rounded large
                :disabled="!valid"
                :loading="loading"
                color="success"
                type="submit"
              >{{ $t('project.explore.edit.submit') }}
              </v-btn>
            </v-col>
          </v-row>
        </v-form>
      </v-container>
    </v-card>
  </div>
</template>

<script lang="ts">
import { Component, Vue } from 'vue-property-decorator'
import Author from '@/models/paperEntities/author'

@Component
export default class AuthorEditPage extends Vue {
  valid: boolean = true
  loading: boolean = false
  firstName: string = ''
  lastName: string = ''

  nameRules: Array<Function> = [
    (v: string | undefined) => !!v || this.$root.$t('project.explore.edit.author.name_required'),
    (v: string | undefined) => (v && v.trim().length > 0) || this.$root.$t('project.explore.edit.author.name_not_empty')
  ]

  urlRules: Array<Function> = []

  get author (): Author {
    return this.$store.getters['paperEntities/activeEntity'] as Author
  }

  beforeMount (): void {
    this.firstName = this.author.firstName
    this.lastName = this.author.lastName
  }

  async submit (): Promise<void> {
    if (!this.valid) return
    this.loading = true
    await this.$store.dispatch('paperEntities/updateEntity', {
      params: {
        projectId: this.$route.params.projectId,
        entityType: this.$route.params.queryByType,
        id: this.$route.params.entityId
      },
      data: {
        firstName: this.firstName,
        lastName: this.lastName
      }
    })
    this.loading = false
    this.$router.back()
  }

  back (): void {
    this.$router.back()
  }
}
</script>
