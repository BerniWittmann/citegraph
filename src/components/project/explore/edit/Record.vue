<template>
  <div class="project-explore-single-record-edit">
    <v-card class="pa-4">
      <v-card-title class="headline">
        <h2 class="display-5" v-once>{{record.title }}</h2>
      </v-card-title>
      <v-container>
        <v-form ref="form" @submit.prevent="submit" class="mt-12" v-model="valid">
          <v-row justify="center">
            <v-col lg="6">
              <v-text-field
                :label="$t('project.explore.edit.record.title')"
                v-model="title"
                :rules="nameRules"
                required
                autofocus/>
            </v-col>
          </v-row>
          <v-row class="my-10" justify="space-between">
            <v-col cols="auto">
              <v-btn rounded large color="warning" @click="back">
                <v-icon>mdi-menu-left</v-icon>
                {{ $t('project.explore.edit.back') }}
              </v-btn>
            </v-col>

            <v-col cols="auto">
              <v-btn
                rounded large
                :disabled="!valid"
                :loading="loading"
                color="success"
                type="submit"
              >{{ $t('project.explore.edit.submit') }}
              </v-btn>
            </v-col>
          </v-row>
        </v-form>
      </v-container>
    </v-card>
  </div>
</template>

<script lang="ts">
import { Component, Vue } from 'vue-property-decorator'
import Record from '@/models/paperEntities/record'

@Component
export default class RecordEditPage extends Vue {
  valid: boolean = true
  loading: boolean = false
  title: string = ''

  nameRules: Array<Function> = [
    (v: string | undefined) => !!v || this.$root.$t('project.explore.edit.record.name_required'),
    (v: string | undefined) => (v && v.trim().length > 0) || this.$root.$t('project.explore.edit.record.name_not_empty')
  ]

  get record (): Record {
    return this.$store.getters['paperEntities/activeEntity'] as Record
  }

  beforeMount (): void {
    this.title = this.record.title
  }

  async submit (): Promise<void> {
    if (!this.valid) return
    this.loading = true
    await this.$store.dispatch('paperEntities/updateEntity', {
      params: {
        projectId: this.$route.params.projectId,
        entityType: this.$route.params.queryByType,
        id: this.$route.params.entityId
      },
      data: {
        title: this.title
      }
    })
    this.loading = false
    this.$router.back()
  }

  back (): void {
    this.$router.back()
  }
}
</script>
