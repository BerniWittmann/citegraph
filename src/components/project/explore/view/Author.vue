<template>
  <div class="project-explore-single-author" v-if="author">
    <layout>
      <template v-slot:title>
        <v-avatar color="primary" class="mr-5" size="62">
          <b class="white--text">{{ initials }}</b>
        </v-avatar>
        <h2 class="display-5">{{ author.firstName }} {{ author.lastName }}</h2>
      </template>
      <v-container>
        <v-row>
          <v-col :xl="3" :lg="6">
            <v-list>
              <v-subheader class="title">{{ $t('project.explore.view.author.metrics') }}</v-subheader>
              <v-list-item>
                <v-list-item-content>{{ $t('project.explore.view.author.count_records') }}:</v-list-item-content>
                <v-list-item-content class="align-end">{{ author.countRecords }}</v-list-item-content>
              </v-list-item>
            </v-list>
          </v-col>
        </v-row>
        <v-row>
          <data-table belongs-to-type="author" :belongs-to="author.id" query-by-type="record"></data-table>
        </v-row>
      </v-container>
    </layout>
  </div>
</template>

<script lang="ts">
import { Component, Vue } from 'vue-property-decorator'
import PaperEntityLayout from '@/layouts/PaperEntity.vue'
import Author from '@/models/paperEntities/author'
import DataTable from '@/components/project/explore/DataTable.vue'

@Component({
  components: {
    DataTable,
    Layout: PaperEntityLayout
  }
})
export default class AuthorViewPage extends Vue {
  get author (): Author {
    return this.$store.getters['paperEntities/activeEntity'] as Author
  }

  get initials (): string {
    return ((this.author.firstName ? this.author.firstName.substring(0, 1) : '') + (this.author.lastName ? this.author.lastName.substring(0, 1) : '')).toUpperCase()
  }
}
</script>
