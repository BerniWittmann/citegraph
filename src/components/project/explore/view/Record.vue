<template>
  <div class="project-explore-single-record">
    <layout>
      <template v-slot:title>
        <h2 class="display-5">{{ record.title }}</h2>
      </template>
      <v-container>
        <v-row>
          <v-col>
            <v-subheader class="title">{{ $t('project.explore.view.record.authors') }}</v-subheader>
            <v-chip-group column>
              <v-chip v-for="(author, index) in record.authors" :key="index"
                      :to="{ name: 'projects.single.explore.view', params: { ...$route.params, queryByType: 'author', entityId: author.id } }">
                {{ author.firstName }} {{ author.lastName }}
              </v-chip>
            </v-chip-group>
          </v-col>
          <v-col :cols="2">
            <v-card class="justify-center text-center pa-2 py-3">
              <b class="display-1 primary--text font-weight-black">{{ record.numberCitations }}</b>
              <p class="headline">{{ $t('project.explore.view.record.citations') }}</p>
            </v-card>
          </v-col>
        </v-row>
        <v-row>
          <v-col>
            <v-subheader class="title">{{ $t('project.explore.view.record.year') }}</v-subheader>
            <p>{{ record.year }}</p>
          </v-col>
        </v-row>
        <v-row>
          <v-col>
            <v-subheader class="title">{{ $t('project.explore.view.record.keywords') }}</v-subheader>
            <v-chip-group column>
              <v-chip v-for="(keyword, index) in record.keywords" small :key="index">{{ keyword }}</v-chip>
            </v-chip-group>
          </v-col>
        </v-row>
      </v-container>
    </layout>
  </div>
</template>

<script lang="ts">
import { Component, Vue } from 'vue-property-decorator'
import PaperEntityLayout from '@/layouts/PaperEntity.vue'
import Record from '@/models/paperEntities/record'

@Component({
  components: {
    Layout: PaperEntityLayout
  }
})
export default class RecordViewPage extends Vue {
  get record (): Record {
    return this.$store.getters['paperEntities/activeEntity'] as Record
  }
}
</script>
