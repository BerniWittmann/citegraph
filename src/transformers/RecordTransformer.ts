/* eslint-disable camelcase */
import Transformer from './Transformer'
import Record from '@/models/paperEntities/record'
import AuthorTransformer, { AuthorDTO } from '@/transformers/AuthorTransformer'

export interface RecordDTO {
  id: string
  title: string;
  authors?: Array<AuthorDTO>
  year: number
  keywords: Array<string>
  number_citations: number
}

export default class RecordTransformer extends Transformer {
  static fetch (item: RecordDTO): Record {
    return new Record({
      id: item.id,
      title: item.title,
      year: item.year,
      numberCitations: item.number_citations,
      keywords: item.keywords,
      authors: item.authors ? AuthorTransformer.fetchCollection(item.authors) : []
    })
  }

  static send (data: Record): RecordDTO {
    return {
      id: data.id,
      title: data.title,
      year: data.year,
      number_citations: data.numberCitations,
      keywords: data.keywords,
      authors: data.authors ? AuthorTransformer.sendCollection(data.authors) : []
    }
  }
}
