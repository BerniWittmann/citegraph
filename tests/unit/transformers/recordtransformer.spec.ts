import Author from '@/models/paperEntities/author'
import RecordTransformer from '@/transformers/RecordTransformer'
import Record from '@/models/paperEntities/record'

describe('transformers/Recordtransformer', () => {
  describe('can send a record', () => {
    it('sends a record', () => {
      expect(RecordTransformer.send(new Record({
        id: '1',
        title: 'First',
        year: 2000,
        numberCitations: 1,
        authors: [],
        keywords: []
      }))).toMatchSnapshot()
    })
    it('sends the authors', () => {
      expect(RecordTransformer.send(new Record({
        id: '1',
        title: 'First',
        year: 2000,
        numberCitations: 1,
        authors: [
          new Author({ id: '1', firstName: 'First', lastName: 'Author', countRecords: 12 })
        ],
        keywords: []
      }))).toMatchSnapshot()
    })
  })

  describe('can fetch a record', () => {
    it('fetches a record', () => {
      expect(RecordTransformer.fetch({
        id: '1',
        title: 'First',
        year: 2000,
        number_citations: 1,
        keywords: []
      })).toMatchSnapshot()
    })
    it('fetches the authors', () => {
      expect(RecordTransformer.fetch({
        id: '1',
        title: 'First',
        year: 2000,
        number_citations: 1,
        authors: [
          { id: '1', first_name: 'First', last_name: 'Author', count_records: 12 }
        ],
        keywords: []
      })).toMatchSnapshot()
    })
  })
})
