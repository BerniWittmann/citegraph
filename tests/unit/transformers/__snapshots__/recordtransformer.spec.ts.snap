// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transformers/Recordtransformer can fetch a record fetches a record 1`] = `
Record {
  "authors": Array [],
  "id": "1",
  "keywords": Array [],
  "numberCitations": 1,
  "title": "First",
  "year": 2000,
}
`;

exports[`transformers/Recordtransformer can fetch a record fetches the authors 1`] = `
Record {
  "authors": Array [
    Author {
      "countRecords": 12,
      "firstName": "First",
      "id": "1",
      "lastName": "Author",
    },
  ],
  "id": "1",
  "keywords": Array [],
  "numberCitations": 1,
  "title": "First",
  "year": 2000,
}
`;

exports[`transformers/Recordtransformer can send a record sends a record 1`] = `
Object {
  "authors": Array [],
  "id": "1",
  "keywords": Array [],
  "number_citations": 1,
  "title": "First",
  "year": 2000,
}
`;

exports[`transformers/Recordtransformer can send a record sends the authors 1`] = `
Object {
  "authors": Array [
    Object {
      "count_records": 12,
      "first_name": "First",
      "id": "1",
      "last_name": "Author",
    },
  ],
  "id": "1",
  "keywords": Array [],
  "number_citations": 1,
  "title": "First",
  "year": 2000,
}
`;
